pkg/wireguard/agent/agent.go

```diff
@@ -0,0 +1,352 @@
+	err := netlink.LinkAdd(link)
+	if err != nil && !errors.Is(err, unix.EEXIST) {
+		return err
+	}
+
+	if err := sysctl.Write(fmt.Sprintf("net.ipv4.conf.%s.rp_filter", IfaceName), "2"); err != nil {
+		return nil
+	}
+
+	type param struct {
+		ip     *net.IPNet
+		family int
+	}
+	params := []param{}
+	if option.Config.EnableIPv4 {
+		ip := &net.IPNet{
+			IP:   node.GetWireguardIPv4(),
+			Mask: a.wireguardV4CIDR.Mask,
+		}
+		params = append(params, param{ip: ip, family: netlink.FAMILY_V4})
+	}
+	if option.Config.EnableIPv6 {
+		ip := &net.IPNet{
+			IP:   node.GetWireguardIPv6(),
+			Mask: a.wireguardV6CIDR.Mask,
+		}
+		params = append(params, param{ip: ip, family: netlink.FAMILY_V6})
+	}
+	for _, p := range params {
+		// Removes stale IP addresses from wg device
+		addrs, err := netlink.AddrList(link, p.family)
+		if err != nil {
+			return err
+		}
+		for _, addr := range addrs {
+			if !cidr.NewCIDR(addr.IPNet).Equal(cidr.NewCIDR(p.ip)) {
+				if err := netlink.AddrDel(link, &addr); err != nil {
+					return fmt.Errorf("failed to remove stale wg ip: %w", err)
+				}
+			}
+		}
+
+		err = netlink.AddrAdd(link, &netlink.Addr{IPNet: p.ip})
+		if err != nil && !errors.Is(err, unix.EEXIST) {
+			return err
+		}
+	}
+
+	cfg := &wgtypes.Config{
+		PrivateKey:   &a.privKey,
+		ListenPort:   &a.listenPort,
+		ReplacePeers: false,
+	}
+	if err := a.wgClient.ConfigureDevice(IfaceName, *cfg); err != nil {
+		return err
+	}
+
+	if err := netlink.LinkSetUp(link); err != nil {
+		return err
+	}
+
+	dev, err := a.wgClient.Device(IfaceName)
+	if err != nil {
+		return err
+	}
+	for _, peer := range dev.Peers {
+		a.restoredPubKeys[peer.PublicKey.String()] = struct{}{}
+	}
+
+	// Create the rule to steer the marked traffic (from a local endpoint to a
+	// remote endpoint) via the Wireguard tunnel
+	rule := route.Rule{
+		Priority: 1,
+		Mark:     linux_defaults.RouteMarkEncrypt,
+		Mask:     linux_defaults.RouteMarkMask,
+		Table:    linux_defaults.RouteTableWireguard,
+	}
+	if err := route.ReplaceRule(rule); err != nil {
+		return err
+	}
+
+	ip, net, _ := net.ParseCIDR("0.0.0.0/0")
+	net.IP = ip
+	rt := route.Route{
+		Prefix: *net,
+		Device: IfaceName,
+		Table:  linux_defaults.RouteTableWireguard,
+	}
+	if _, err := route.Upsert(rt); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a *Agent) RestoreFinished() error {
+	a.Lock()
+	defer a.Unlock()
+
+	// Delete obsolete peers
+	for _, pubKeyHex := range a.pubKeyByNodeName {
+		delete(a.restoredPubKeys, pubKeyHex)
+	}
+	for pubKeyHex := range a.restoredPubKeys {
+		log.WithField("pubKey", pubKeyHex).Info("Removing obsolete peer")
+		if err := a.deletePeerByPubKey(pubKeyHex); err != nil {
+			return err
+		}
+	}
+
+	a.restoredPubKeys = nil
+
+	log.Info("Finished restore")
+
+	return nil
+}
+
+func (a *Agent) UpdatePeer(nodeName, pubKeyHex string,
+	wgIPv4, nodeIPv4 net.IP, podCIDRv4 *net.IPNet,
+	wgIPv6, nodeIPv6 net.IP, podCIDRv6 *net.IPNet) error {
```

Comment:
Should we define a `Peer` type? Would that make the code more flexible in the future or do you feel it's not necessary?

Reply:
Martynas and I discussed this, and we'll think it's probably better to leave it as is for now. We will likely extend this method in a follow-up to support managed Kubernetes. Once we do that, we'll have a better understanding of how the `Peer` struct needs to look like.